### PR TITLE
#1243: Fixed forward function overloads resolving incorrectly

### PR DESCRIPTION
**Pull request by brynrhodes**

### CODE REVIEW

File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/FunctionDefinitionInfo.java [hunk 1]
```diff
@@ -19,12 +19,12 @@ public void setName(String value) {
         name = value;
     }
 
-    public void setPreCompileOutput(PreCompileOutput preCompileOutput) {
-        this.preCompileOutput = preCompileOutput;
+    public void setPreCompileOutput(FunctionHeader functionHeader) {
```
Comment (JPercival) [inline Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/FunctionDefinitionInfo.java:22]: Suggestion: This should be get/setFunctionHeader

File: Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/fhir/r401/TestOverloadForward.cql [hunk 2]
```diff
@@ -0,0 +1,18 @@
+library TestOverloadForward
+
+using FHIR version '4.0.1'
```
Comment (JPercival) [inline Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/fhir/r401/TestOverloadForward.cql:3]: Question: Do we need to consider the case of header-only declarations for an included library? I _think_ given the current compiler architecture it's expected that an included Library will be fully compiled but I'm not 100% certain.

File: Src/java/cql-to-elm/src/test/java/org/cqframework/cql/cql2elm/fhir/r401/BaseTest.java [hunk 3]
```diff
@@ -831,4 +832,30 @@ public void testInclude() throws IOException {
         assertThat(qlr.getName(), equalTo("M"));
         assertThat(eqv.getOperand().get(1), instanceOf(ToConcept.class));
     }
+
+    @Test
+    public void testOverload() throws IOException {
+        CqlTranslator translator = TestUtils.runSemanticTest("fhir/r401/TestOverload.cql", 0);
```
Comment (JPercival) [inline Src/java/cql-to-elm/src/test/java/org/cqframework/cql/cql2elm/fhir/r401/BaseTest.java:838]: Suggestion: We should extend the tests to ensure that all functions are compiled, even if they aren't referenced. We now compile on-demand, so there may be dangling definitions that would "disappear" from the ELM.

**Review by JPercival — APPROVED** [review]
Approved with some questions/suggestions.

File: Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/fhir/r401/TestOverloadForward.cql (hunk 2 above)
Comment (brynrhodes) [reply Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/fhir/r401/TestOverloadForward.cql:3]: Yes, the expectation is that included libraries are fully compiled and the current translator will ensure that is the case. I tried to do lazy compilation even for included libraries, but it proved to be more trouble than it was worth at this point.

File: Src/java/cql-to-elm/src/test/java/org/cqframework/cql/cql2elm/fhir/r401/BaseTest.java (hunk 3 above)
Comment (brynrhodes) [reply Src/java/cql-to-elm/src/test/java/org/cqframework/cql/cql2elm/fhir/r401/BaseTest.java:838]: We could construct a specific test, but I definitely encountered this issue in getting past the regression tests, I initially tried to leave some functions uncompiled, but reconstructing the state necessary to do nested compilation was too big a lift and I opted for ensuring the Library is 100% compiled instead.

**Review by mdnazmulkarim — APPROVED** [review]

**Pull request merged**